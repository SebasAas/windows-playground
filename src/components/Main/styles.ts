import styled from 'styled-components';

export const Wrapper = styled.main`
  display: flex;
  flex-wrap: wrap;
  flex-direction: column;
  justify-content: flex-start;
  align-content: flex-start;
  height: 100vh;
  width: 100vw;
  position: relative;
  background-image: url('img/windows-background.png');
  background-repeat: no-repeat;
  background-size: cover;
  background-position: center;
  overflow: hidden;
`;

export const Button = styled.button`
  height: auto;
  width: 100%;
  background-color: #ececec;
  border: none;
  border-bottom: 1px solid #c4c4c4;
  cursor: pointer;
  color: #3e3e3e;
  font-size: 14px;
  padding: 10px;

  &:hover {
    background-color: #fff;
  }
`;

export const CreateFolderWrapper = styled.div`
  height: calc(100% - 35px);
  display: flex;
  justify-content: flex-start;
  align-items: center;
  flex-direction: column;

  h1 {
    font-size: 28px;
    font-weight: 500;
    margin: 20px 0;
  }

  input {
    height: 30px;
    width: 45%;
    text-align: center;
    border: none;
    border-bottom: 1px solid;
    color: #505050;
    margin-bottom: 20px;
  }

  button {
    margin-bottom: 20px;
    height: 30px;
    width: 90px;
    background: transparent;
    border-radius: 5px;
    background-color: #52339e;
    color: white;
    border-color: transparent;
    cursor: pointer;
  }
`;
