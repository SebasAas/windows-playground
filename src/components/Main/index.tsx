import { useState, useEffect, useRef } from 'react';

// Components
import Modal from 'components/Modal';

// Styles
import * as S from './styles';

// Store
import { RootState } from 'store/store';
import { useSelector, useDispatch } from 'react-redux';
import {
  createFile,
  ApplicationsTypes
} from 'store/features/applications/applicationsSlice';
import {
  ModalTypes,
  openModalOption,
  closeModalOption,
  openModalCreateFolder,
  closeModalCreateFolder
} from 'store/features/modal/modalSlice';

// Hooks
import useOnClickOutside from 'hooks/useOnClickOutside';

type PropsMain = {
  children?: JSX.Element | JSX.Element[];
};

function Main({ children }: PropsMain) {
  const dispatch = useDispatch();
  const modalRef = useRef<HTMLElement>(null);
  const applications: ApplicationsTypes[] = useSelector(
    (state: RootState) => state.applications
  );
  const modal: ModalTypes = useSelector((state: RootState) => state.modal);

  const {
    showModalOption: isModalOptionOpen,
    showModalCreateFolder: isModalCreateFolderOpen
  } = modal;

  const [folderName, setFolderName] = useState('');
  const [coordinates, setCoordinates] = useState({ posx: 0, posy: 0 });

  useEffect(() => {
    window.addEventListener('contextmenu', (e) => handleModalOption(e));
    // window.addEventListener('click', () => closeModalOption());
    return () => {
      window.removeEventListener('contextmenu', (e) => handleModalOption(e));
      // window.removeEventListener('click', () => closeModalOption());
    };
  }, []);

  const handleModalOption = (e: any) => {
    setCoordinates({ posx: e.clientX, posy: e.clientY });
    dispatch(openModalOption());
  };

  const handleCreateFolder = () => {
    dispatch(closeModalOption());
    dispatch(openModalCreateFolder());
  };

  const handleCreateObject = () => {
    const obj = {
      id: applications[applications.length - 1].id + 1,
      name: folderName,
      type: 'folder',
      image: '/img/folder.png',
      location: 'desktop',
      options: {
        show: false,
        minimize: false,
        maximize: false,
        details: {
          files: []
        }
      }
    };

    dispatch(createFile(obj));
  };

  useOnClickOutside(modalRef, closeModalOption(), 'MAIN');

  const handleModalOptions = () => {
    if (isModalOptionOpen)
      return (
        <Modal
          draggable={false}
          coordinates={coordinates}
          onClose={() => dispatch(closeModalOption())}
          innerRef={modalRef}
        >
          <S.Button onClick={() => handleCreateFolder()}>
            Create Folder
          </S.Button>
          <S.Button>Properties</S.Button>
        </Modal>
      );
  };

  const handleModalCreateFolder = () => {
    if (isModalCreateFolderOpen)
      return (
        <Modal
          coordinates={{ posy: 150, posx: 400 }}
          size={{ height: 'auto', width: '400px' }}
          onClose={() => dispatch(closeModalCreateFolder())}
        >
          <S.CreateFolderWrapper>
            <h1>Create Folder</h1>
            <input
              type="text"
              placeholder="Folder Name"
              value={folderName}
              onChange={(e) => setFolderName(e.target.value)}
            />
            <button onClick={() => handleCreateObject()}>Create</button>
          </S.CreateFolderWrapper>
        </Modal>
      );
  };

  return (
    <S.Wrapper>
      {handleModalOptions()}
      {handleModalCreateFolder()}
      {children}
    </S.Wrapper>
  );
}

export default Main;
