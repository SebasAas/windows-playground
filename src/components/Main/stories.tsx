import Main from '.';
import { ComponentStory, ComponentMeta } from '@storybook/react';

export default {
  title: 'Main',
  component: Main
} as ComponentMeta<typeof Main>;

const Template: ComponentStory<typeof Main> = () => <Main />;

export const Basic = Template.bind({});
Basic.args = {};
