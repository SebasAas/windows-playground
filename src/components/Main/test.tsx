import { render, screen } from '@testing-library/react';

import Main from '.';

describe('Should render a component Main', () => {
  it('should render a component', () => {
    const { container } = render(<Main />);

    // expect(
    //   screen.getByRole('heading', { name: /advanced react/i })
    // ).toBeInTheDocument();

    expect(container.firstChild).toMatchSnapshot();
  });

  it('should render a component with styles', () => {
    const { container } = render(<Main />);

    console.log(container);
    // expect(container.firstChild).toHaveStyle({ 'background-color': '#06092b' });
  });
});
