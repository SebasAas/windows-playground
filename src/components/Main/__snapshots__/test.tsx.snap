// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should render a component Main should render a component 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  height: 100vh;
  width: 100vw;
  position: relative;
  background-image: url('img/windows-background.png');
  background-repeat: no-repeat;
  background-size: cover;
  background-position: center;
  overflow: hidden;
}

<main
  class="c0"
/>
`;
